tests/unit/test_download.py

```diff
@@ -339,3 +340,98 @@ def test_parse_credentials():
+    get = functools.partial(
+        auth._get_url_and_credentials,
+        allow_netrc=False,
+        allow_keyring=False
+    )
+
+    # Check URL parsing
+    assert get("http://foo:bar@example.com/path") \
+        == ('http://example.com/path', 'foo', 'bar')
+    assert auth.passwords['example.com'] == ('foo', 'bar')
+
+    auth.passwords['example.com'] = ('user', 'pass')
+    assert get("http://foo:bar@example.com/path") \
+        == ('http://example.com/path', 'user', 'pass')
+
+
+def test_get_credentials_keyring():
+    try:
+        import keyring
+        import keyring.backend
+        import keyring.core
+    except ImportError:
+        pytest.skip("keyring required for test")
+
+    auth = MultiDomainBasicAuth()
+    get = functools.partial(
+        auth._get_url_and_credentials,
+        allow_netrc=False,
+        allow_keyring=True
+    )
+
+    # Specify a backend with known responses for specific systems.
+    # Validates that we (a) pass through the correct system, and
+    # (b) pass back the correct result.
+    class GetPasswordBackendV1(object):
+        def get_password(self, system, username):
+            if system == "http://example.com/path1":
+                return (username or "") + "!password1"
+            return None
+
+    class GetPasswordBackendV2(GetPasswordBackendV1):
+        def get_credential(self, system, username):
+            if system == "http://example.com/path2":
+                return keyring.backend.Credential("username", "password2")
+            return None
+
+    keyring.backend.KeyringBackend.register(GetPasswordBackendV1)
+
+    test_keyring1 = GetPasswordBackendV1()
+    test_keyring2 = GetPasswordBackendV2()
+    orig_keyring = keyring.core.get_keyring()
+    try:
+        # Enable the old-style keyring backend. This does *not*
+        # support adding or overriding the username.
+        keyring.core.set_keyring(test_keyring1)
+
+        auth.passwords.pop('example.com', None)
+        assert get("http://example.com/path1") \
+            == ('http://example.com/path1', None, None)
+        auth.passwords.pop('example.com', None)
+
+        assert get("http://user@example.com/path1") \
+            == ('http://example.com/path1', "user", "user!password1")
+        auth.passwords.pop('example.com', None)
+
+        assert get("http://user2@example.com/path1") \
+            == ('http://example.com/path1', "user2", "user2!password1")
+        auth.passwords.pop('example.com', None)
+
+        assert get("http://example.com/path2") \
+            == ('http://example.com/path2', None, None)
+        auth.passwords.pop('example.com', None)
+
+        assert get("http://foo@example.com/path2") \
+            == ('http://example.com/path2', "foo", None)
+        auth.passwords.pop('example.com', None)
+
+        # Enable the get_username_and_password API. This allows
+        # keyring backends to provide or override the username.
+        keyring.core.set_keyring(test_keyring2)
```

Comment:
It looks like it could help to break this up into two tests, one for each backend (probably with a helper function and/or context manager for the common setup). The test is pretty long.